refactor: certificates generation package and secrets

fabriziosestito opened the pull request:
## Description
This PR refactors the generation utilities in the `admissionregistration` package,
removing duplication and simplifying the current implementation.
It renames `admissionregistration` in `certs` and reactors the existing reconciler unit test in integration tests clauses.

## Additional info

- This PR also changes the golangci configuration to add an exception for the `ok` map index in the `varnamelen` linter. However we should uniform and clean up the golangci configurations across all the KW repos, this is tracked by: #778 

- There are a few incongruences with the naming, for instance, this status  https://github.com/kubewarden/kubewarden-controller/blob/030edce6f7c57d54b88453054e39f9baa20377cc/pkg/apis/policies/v1/policyserver_types.go#L132 should be probably named `CertSecretReconciled` instead. Fixing this is out of scope for this PR and I will create an issue about it.

Fixes: #775, #664

flavio commented on controllers/utils_test.go:
```diff
@@ -164,6 +164,14 @@ func getTestPolicyServerService(policyServerName string) (*corev1.Service, error
 	return &service, nil
 }
 
+func getTestPolicyServerCASecret() (*corev1.Secret, error) {
```
This is retrieving the Root CA created by our controller. I suggest to rename this helper function to reflect that.

Maybe something like `getInternalRootCASecret` ?

flavio commented on internal/pkg/admission/policy-server-ca-secret.go:
```diff
@@ -12,21 +12,18 @@ import (
-type generateCertFunc = func(ca []byte, commonName string, extraSANs []string, CAPrivateKey *rsa.PrivateKey) ([]byte, []byte, error)
-
-func (r *Reconciler) fetchOrInitializePolicyServerCASecret(ctx context.Context, policyServer *policiesv1.PolicyServer, caSecret *corev1.Secret, generateCert generateCertFunc) error {
+func (r *Reconciler) fetchOrInitializePolicyServerCASecret(ctx context.Context, policyServer *policiesv1.PolicyServer, caSecret *corev1.Secret) error {
```
Again, there's a bit of naming confusion here we inherited from the past.

This method creates the certificate used by a specific Policy Server instance, there's no CA associated with a Policy Server

flavio commented on internal/pkg/admission/policy-server-ca-secret.go:
```diff
@@ -73,37 +83,40 @@ func (r *Reconciler) fetchOrInitializePolicyServerCASecret(ctx context.Context,
 }
 
-func (r *Reconciler) fetchOrInitializePolicyServerCARootSecret(ctx context.Context, policyServer *policiesv1.PolicyServer, generateCA generateCAFunc, pemEncodeCertificate pemEncodeCertificateFunc) (*corev1.Secret, error) {
+func (r *Reconciler) fetchOrInitializePolicyServerCARootSecret(ctx context.Context, policyServer *policiesv1.PolicyServer) (*corev1.Secret, error) {
```
Same here, the naming is confusing since no CA is stored inside of the `Secret` that is returned

flavio commented on internal/pkg/admission/policy-server-ca-secret.go:
```diff
@@ -115,29 +128,34 @@ func (r *Reconciler) fetchOrInitializePolicyServerCARootSecret(ctx context.Conte
 
-func updateSecretCA(policyServerSecret *corev1.Secret, generateCA generateCAFunc, pemEncodeCertificate pemEncodeCertificateFunc) error {
-	caRoot, err := generateCA()
+func updateCASecret(policyServerSecret *corev1.Secret) error {
```
This function is always generating the CA, it doesn't look into the contents of the secret that is passed as parameter.

Why don't we change the signature of the function to be something like that?

```golang
func createInternalCASecret() (policyServerSecret *corev1.Secret, error)
```

flavio commented on internal/pkg/admission/policy-server-ca-secret.go:
Maybe something for later, but I would also rename this file since it can mislead into thinking each Policy Server has a dedicated CA

flavio commented on internal/pkg/certs/ca.go:
```diff
@@ -10,21 +10,18 @@ import (
-
-func GenerateCA() (*CA, error) {
-	privateKey, err := newPrivateKey(1024)
+func GenerateCA() ([]byte, *rsa.PrivateKey, error) {
```
Can you add some inline documentation explaining what is returned by the function?

flavio commented on internal/pkg/certs/cert.go:
```diff
@@ -30,7 +28,7 @@ func GenerateCert(ca []byte,
 	// key size must be higher than 1024, otherwise the PolicyServer
 	// TLS acceptor will refuse to start
-	servingPrivateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	privateKey, err := rsa.GenerateKey(rand.Reader, 4096)
```
Maybe we could create a constant defining the size of the key? This would be used to create the internal CA and the Policy Server certs.

flavio commented on internal/pkg/certs/cert.go:
```diff
@@ -16,8 +14,8 @@ import (
 func GenerateCert(ca []byte,
```
Please add some inline docs

flavio reviewed (changes requested):
I think we should address some naming confusion about Policy Server CA, see my notes.

Overall LGTM

fabriziosestito commented:
@flavio I've addressed the comments in https://github.com/kubewarden/kubewarden-controller/pull/777/commits/c3cbdf4e9759e6e0fa2900d2343bcde0865cfe43. 

I tracked all the naming-related comments in: https://github.com/kubewarden/kubewarden-controller/issues/782

fabriziosestito commented:
Blocked: waiting for 1.14 release

flavio reviewed (approved):
Approving, the naming changes I've suggested are going to be done inside of future PRs
